MapleServer2/Types/Inventory.cs

```diff
@@ -6,73 +6,131 @@
+        {
             this.Size = size;
             this.items = new Dictionary<long, Item>();
             byte maxTabs = Enum.GetValues(typeof(InventoryTab)).Cast<byte>().Max();
             this.slotMaps = new Dictionary<short, long>[maxTabs + 1];
-            for (byte i = 0; i <= maxTabs; i++) {
+            for (byte i = 0; i <= maxTabs; i++)
+            {
                 this.slotMaps[i] = new Dictionary<short, long>();
             }
         }
 
-        public Inventory(short size, IEnumerable<Item> loadItems) : this(size) {
-            foreach (Item item in loadItems) {
+        public Inventory(short size, IEnumerable<Item> loadItems) : this(size)
+        {
+            foreach (Item item in loadItems)
+            {
                 Add(item);
             }
         }
 
-        public ICollection<Item> GetItems(InventoryTab tab) {
+        public ICollection<Item> GetItems(InventoryTab tab)
+        {
             return GetSlots(tab).Select(kvp => items[kvp.Value])
                 .ToImmutableList();
         }
 
         // TODO: precompute next free slot to avoid iteration on Add
         // TODO: Stack items when they are the same
         // Returns false if inventory is full
-        public bool Add(Item item) {
+        public bool Add(Item item)
+        {
             // Item has a slot set, try to use that slot
-            if (item.Slot >= 0) {
-                if (!SlotTaken(item, item.Slot)) {
-                    AddInternal(item);
+            if (item.Slot >= 0)
+            {
+                if (!SlotTaken(item, item.Slot))
+                {
+                    if (isStackable(item))
+                    {
+                        Stack(item);
+                    }
+                    else
+                    {
+                        AddInternal(item);
+                    }
                     return true;
                 }
 
                 item.Slot = -1; // Reset slot
             }
 
-            for (short i = 0; i < Size; i++) {
-                if (!SlotTaken(item, i)) {
+            for (short i = 0; i < Size; i++)
+            {
+                if (!SlotTaken(item, i))
+                {
                     item.Slot = i;
-                    AddInternal(item);
+                    if (isStackable(item))
+                    {
+                        Stack(item);
+                    }
+                    else
+                    {
+                        AddInternal(item);
+                    }
                     return true;
                 }
             }
-
             return false;
         }
 
+        // Checks if item is stackable.
+        private bool isStackable(Item item)
+        {
+            return item.SlotMax > 1;
+        }
+
+        private void Stack(Item item) // Replace 10 with item.slotmax to test against actual max slot count.
+        {
+            foreach (Item i in items.Values)
+            {
+                if (i.Id == item.Id)
```

Comment:
Can you invert this condition and if this new condition is true, add a `continue;`?